python/adjoint/objective.py

```diff
@@ -231,98 +231,52 @@ def __call__(self):
 
 
 class FourierFields(ObjectiveQuantity):
-    def __init__(self, sim, volume, component, decimation_factor=0):
+    def __init__(self, sim, volume, component, yee_grid=False, decimation_factor=0):
         super().__init__(sim)
-        self.volume = volume
+        self.volume = sim._fit_volume_to_simulation(volume)
         self.component = component
+        self.yee_grid = yee_grid
         self.decimation_factor = decimation_factor
 
     def register_monitors(self, frequencies):
         self._frequencies = np.asarray(frequencies)
         self._monitor = self.sim.add_dft_fields(
-            [self.component],
-            self._frequencies,
-            where=self.volume,
-            yee_grid=False,
-            decimation_factor=self.decimation_factor,
-        )
+            [self.component], self._frequencies, where=self.volume, yee_grid=self.yee_grid, decimation_factor=self.decimation_factor)
         return self._monitor
 
     def place_adjoint_source(self, dJ):
         time_src = self._create_time_profile()
         sources = []
-        scale = self._adj_src_scale()
 
-        x_dim, y_dim, z_dim = len(self._dg.x), len(self._dg.y), len(self._dg.z)
-
-        if self.num_freq == 1:
-            amp = -dJ[0].copy().reshape(x_dim, y_dim, z_dim) * scale
-            if self.component in [mp.Hx, mp.Hy, mp.Hz]:
-                amp = -amp
-            for zi in range(z_dim):
-                for yi in range(y_dim):
-                    for xi in range(x_dim):
-                        if amp[xi, yi, zi] != 0:
-                            sources += [
-                                mp.Source(
-                                    time_src,
-                                    component=self.component,
-                                    amplitude=amp[xi, yi, zi],
-                                    center=mp.Vector3(self._dg.x[xi],
-                                                      self._dg.y[yi],
-                                                      self._dg.z[zi]),
-                                )
-                            ]
+        thickness_indicator = self.sim.fields.indicate_thick_dims(self.volume.swigobj)
+        min_max_corners = self.sim.fields.get_corners(self._monitor.swigobj, self.component) # ivec values of the corners
+        raw_size = (min_max_corners[3:6]-min_max_corners[0:3])/2+np.ones(3)
+        true_size = np.array([raw_size[i]*(thickness_indicator[i]==1)+(thickness_indicator[i]==0) for i in range(3)],dtype=int) # monitor size
```

Comment:
Perhaps a name like `_dft_monitor_size` sounds more pertinent. The function would be
```
    def _dft_monitor_size(self,min_max_corners):
        thickness_indicator = self.sim.fields.indicate_thick_dims(self.volume.swigobj)
        raw_size = (min_max_corners[3:6]-min_max_corners[0:3])/2+np.ones(3)
        true_size = np.array([raw_size[i]*(thickness_indicator[i]==1)+(thickness_indicator[i]==0) for i in range(3)],dtype=int)
        return true_size
```
Line 252, which computes `min_max_corners`, is not encapsulated here, because `min_max_corners` is needed by `fourier_sourcedata` later.